milatools/cli/local.py

```diff
@@ -1,65 +1,92 @@
+from __future__ import annotations
+
 import subprocess
+from subprocess import CompletedProcess
+from typing import Callable, Iterable, Sequence
+
+from typing_extensions import Concatenate, ParamSpec
 
 from .utils import CommandNotFoundError, T, shjoin
 
+P = ParamSpec("P")
+Args = Concatenate[Sequence[str], P]
+
 
 class Local:
-    def display(self, args):
-        print(T.bold_green(f"(local) $ ", shjoin(args)))
+    def display(self, args: Iterable[str]) -> None:
+        print(T.bold_green("(local) $ ", shjoin(args)))
 
-    def silent_get(self, *args, **kwargs):
-        return subprocess.check_output(
-            args,
-            universal_newlines=True,
-            **kwargs,
-        )
+    def silent_get(
+        self,
+        cmd: Sequence[str],
+        _check_output_fn: Callable[
+            Concatenate[str | Sequence[str], P], str
+        ] = subprocess.check_output,
+        *args: P.args,
+        **kwargs: P.kwargs,
+    ) -> str:
+        kwargs["universal_newlines"] = True
+        return _check_output_fn(cmd, *args, **kwargs)
 
-    def get(self, *args, **kwargs):
-        self.display(args)
-        return subprocess.check_output(
-            args,
-            universal_newlines=True,
-            **kwargs,
-        )
+    def get(
+        self,
+        cmd: Sequence[str],
+        _check_output_fn: Callable[
+            Concatenate[str | Sequence[str], P], str
+        ] = subprocess.check_output,
+        *args: P.args,
+        **kwargs: P.kwargs,
+    ):
+        self.display(cmd)
+        kwargs["universal_newlines"] = True
+        return _check_output_fn(cmd, *args, **kwargs)
 
-    def run(self, *args, **kwargs):
-        self.display(args)
+    def run(
+        self,
+        cmd: Sequence[str],
+        _run_fn: Callable[Args[P], CompletedProcess[str]] = subprocess.run,
+        *args: P.args,
+        **kwargs: P.kwargs,
+    ) -> CompletedProcess[str]:
```

Comment:
`*args` should come before `_run_fn` otherwise `run([cmd ...], "arg1", "arg2", {"kwarg3":"kwarg3"})` will result in

```python
cmd = [cmd...]
_run_fn = "arg1"
*args = ("arg2",)
**kwarg3 = {"kwarg3":"kwarg3"}
```

```suggestion
    def run(
        self,
        cmd: Sequence[str],
        *args: P.args,
        _run_fn: Callable[Args[P], CompletedProcess[str]] = subprocess.run,
        **kwargs: P.kwargs,
    ) -> CompletedProcess[str]:
```

Reply:
I'll fix the ordering of arguments, but the explanation for the *args and **kwargs and paramspec is the same as in https://github.com/mila-iqia/milatools/pull/61/files#r1362819307

Reply:
A few notes:
- The wrapped function needs to appear before the *args and **kwargs for us to be able to mimic it's signature with ParamSpec for our `*args` or `**kwargs`.
- You'd first get a type-checking error if you tried to do `run([cmd ...], "arg1", "arg2", {"kwarg3":"kwarg3"})` , so you'd expect to get an error in this case, which is good enough in my opinion.


Assuming that we agree that we want these kinds of methods to be properly typed, we have two options as far as I know:
- Keeping this ParamSpec as-is
- List the arguments of the wrapped function explicitly

The second option would duplicate some of the code and the types of the signature of the wrapped function, but might be more explicit.

Let me know what you think.

Reply:
Ok I opted for the second option in this case, fixed this in 1a6ecd5aa20d545d41807a0e748137353c41a1dd